core/services/head_tracker.go

```diff
@@ -84,17 +79,23 @@ func (ht *HeadTracker) Start() error {
 
 // Stop unsubscribes all connections and fires Disconnect.
 func (ht *HeadTracker) Stop() error {
-	ht.bootMutex.Lock()
-	defer ht.bootMutex.Unlock()
+	ht.headMutex.Lock()
 
 	if !ht.started {
+		ht.headMutex.Unlock()
 		return nil
 	}
 
+	if ht.connected {
+		ht.disconnect()
+		ht.connected = false
+	}
 	close(ht.done)
-	ht.listenForNewHeadsWg.Wait()
 	close(ht.subscriptionSucceeded)
 	ht.started = false
+	ht.headMutex.Unlock()
+
+	ht.listenForNewHeadsWg.Wait()
```

Comment:
What are your reasons for pulling the `listenForNewHeadsWg.Wait()` outside of the mutex?

Reply:
I simplified the head tracker to remove the boot mutex, if the wait happens before the unlock, it causes a deadlock due to contention over the single mutex now used.